### PR TITLE
175 duplicated categories bugfix

### PR DESCRIPTION
**Pull request by goreck888**
fix problems with update and create categories and services

### CODE REVIEW

**Comment by goreck888** [comment]
Fix [#188]

File: db/seeds.rb [hunk 1]
```diff
@@ -33,5 +30,12 @@
 all_providers << Provider.create_with(name: "Provider 3").find_or_create_by(name: "Provider 3")
 all_providers << Provider.create_with(name: "Provider 4").find_or_create_by(name: "Provider 4")
 
-services_size = ENV["services_size"].to_i || 0
-Rake::Task["dev:prime"].invoke(services_size)
+
+puts "Generating services from yaml"
+yaml_hash["services"].each do |service, hash|
+  current = Service.create_with(title: hash["title"], tagline: hash["tagline"], description: hash["description"], provider: Provider.all.sample).find_or_create_by(title: hash["title"])
```
Comment (martaswiatkowska) [inline db/seeds.rb]: use enter

File: db/seeds.rb [hunk 2]
```diff
@@ -33,5 +30,12 @@
...
+  current.categories << Category.find_by(name: hash["parent"])
```
Comment (martaswiatkowska) [inline db/seeds.rb]: hash["parent"]? How it's working?

File: db/seeds.rb [hunk 3]
```diff
@@ -33,5 +30,12 @@
...
+  current.set_first_category_as_main!
+  puts "Generated service #{ hash["title"] }"
+
```
Comment (martaswiatkowska) [inline db/seeds.rb]: I think you can delete this blank line

File: db/data.yml [hunk 4]
```diff
@@ -2,7 +2,10 @@ categories:
 
     Compute:
         name: &computeParent Compute
-        description: General computing services including VM management, containers and job processing
+        description: General computing services including VM management, containers and job processing.
```
Comment (mkasztelnik) [inline db/data.yml]: It looks like you are doing 2 things in one PR
  1. Duplicated categories fix (and duplicated services fix - so this is third thing :))
  1. Adding/Updating `yaml` with additional service information

**Comment by mkasztelnik** [comment]
There is one additional thing I don't like. Every time you are doing `ModelName.create_with(...)` db query e.g.:

```
[18] pry(main)> a=Category.create_with(description: 'asdf')
  Category Load (0.4ms)  SELECT "categories".* FROM "categories"
```
is executed. This can be fixed by changing code into:

```ruby
require "yaml"
puts "Generating categories from yaml"
yaml_hash = YAML.load_file("db/data.yml")

yaml_hash["categories"].each do |_, hash|
  Category.find_or_initialize_by(name: hash["name"]) do |category|
    category.update_attributes(description: hash["description"],
                               parent: Category.find_by(name: hash["parent"]))
  end
  puts "Generated category #{ hash["name"] }"
end

puts "Generating providers"
providers = (1..4).map { |i| Provider.find_or_create_by(name: "Provider #{i}") }


puts "Generating services from yaml"
yaml_hash["services"].each do |_, hash|
  Service.find_or_create_by(title: hash["title"]) do |service|
    service.update_attributes(tagline: hash["tagline"],
                              description: hash["description"],
                              provider: providers.sample,
                              categories: [Category.find_by(name: hash["parent"])])

  end
  puts "Generated service #{ hash["title"] }"
end
```

As you can see I also removed `current.set_first_category_as_main!` line. This is not needed since Service will guard categories and will set first category into main when needed.

This solution has one additional plus: when anything is updated in the category or service except name than it will be updated. This was not a case in the previous version.

**Comment by mkasztelnik** [comment]
According to https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-update-21 `update_attributes!` can be changed into `update!`

File: db/seeds.rb (hunk 2 above)
Comment (goreck888) [reply db/seeds.rb]: It finds category by name included in hash parent in yaml. It works exactly like you can see in the code.

**Comment by mkasztelnik** [comment]
One small remark: changelog entry is missing!

**Pull request merged**